File: gateway_unified/src/claude_gateway/web_search.py
import asyncio
import html
import re
from typing import Any, Dict, Iterable, List
from urllib.parse import parse_qs, unquote, urlparse

import httpx

_RESULT_ANCHOR_RE = re.compile(
    r'<a[^>]*class="result__a"[^>]*href="([^"]+)"[^>]*>(.*?)</a>',
    flags=re.IGNORECASE | re.DOTALL,
)
_RESULT_SNIPPET_RE = re.compile(
    r'class="result__snippet"[^>]*>(.*?)</(?:a|div)>',
    flags=re.IGNORECASE | re.DOTALL,
)
_TAG_RE = re.compile(r"<[^>]+>")
RETRYABLE_SEARCH_ERRORS = (
    httpx.ConnectTimeout,
    httpx.ReadTimeout,
    httpx.ConnectError,
    httpx.RemoteProtocolError,
)


def _strip_tags(raw: str) -> str:
    text = _TAG_RE.sub("", raw)
    return html.unescape(text).strip()


def _normalize_url(raw_href: str) -> str:
    value = html.unescape((raw_href or "").strip())
    if not value:
        return ""

    if value.startswith("//"):
        value = "https:" + value
    elif value.startswith("/l/?"):
        value = "https://duckduckgo.com" + value

    parsed = urlparse(value)
    if "duckduckgo.com" in parsed.netloc and parsed.path == "/l/":
        q = parse_qs(parsed.query)
        uddg = q.get("uddg", [""])[0]
        if uddg:
            return unquote(uddg).strip()
    return value


def _domain_allowed(url: str, allowed_domains: Iterable[str] | None) -> bool:
    if not allowed_domains:
        return True
    netloc = urlparse(url).netloc.lower()
    if not netloc:
        return False
    for item in allowed_domains:
        domain = (item or "").strip().lower()
        if not domain:
            continue
        if netloc == domain or netloc.endswith("." + domain):
            return True
    return False


async def _search_duckduckgo_html_once(
    query: str,
    *,
    max_results: int = 5,
    timeout_s: float = 20.0,
    allowed_domains: Iterable[str] | None = None,
) -> List[Dict[str, str]]:
    q = (query or "").strip()
    if not q:
        return []

    headers = {
        "user-agent": (
            "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
            "AppleWebKit/537.36 (KHTML, like Gecko) Chrome/136.0.0.0 Safari/537.36"
        ),
        "accept-language": "zh-CN,zh;q=0.9,en;q=0.8",
    }
    async with httpx.AsyncClient(follow_redirects=True, timeout=timeout_s) as client:
        resp = await client.get(
            "https://html.duckduckgo.com/html/",
            params={"q": q},
            headers=headers,
        )

    if resp.status_code >= 400:
        return []

    html_text = resp.text
    results: List[Dict[str, str]] = []
    for match in _RESULT_ANCHOR_RE.finditer(html_text):
        href = _normalize_url(match.group(1))
        if not href:
            continue
        if not _domain_allowed(href, allowed_domains):
            continue

        title = _strip_tags(match.group(2))
        if not title:
            continue

        tail = html_text[match.end(): match.end() + 2200]
        snippet_match = _RESULT_SNIPPET_RE.search(tail)
        snippet = _strip_tags(snippet_match.group(1)) if snippet_match else ""

        results.append(
            {
                "title": title,
                "url": href,
                "snippet": snippet,
            }
        )
        if len(results) >= max_results:
            break

    return results


async def search_duckduckgo_html(
    query: str,
    *,
    max_results: int = 5,
    timeout_s: float = 20.0,
    allowed_domains: Iterable[str] | None = None,
    retries: int = 1,
) -> List[Dict[str, str]]:
    last_exc: Exception | None = None
    total_attempts = max(1, retries + 1)
    for attempt in range(total_attempts):
        try:
            return await _search_duckduckgo_html_once(
                query,
                max_results=max_results,
                timeout_s=timeout_s,
                allowed_domains=allowed_domains,
            )
        except RETRYABLE_SEARCH_ERRORS as exc:
            last_exc = exc
            print(
                "[gateway web_search] retryable_error "
                f"attempt={attempt + 1}/{total_attempts} "
                f"type={type(exc).__name__}"
            )
            if attempt >= total_attempts - 1:
                raise
            await asyncio.sleep(0.35)

    if last_exc is not None:
        raise last_exc
    return []


def format_web_search_tool_result_text(query: str, results: List[Dict[str, Any]]) -> str:
    q = (query or "").strip()
    lines = [f"query: {q or '(empty)'}"]
    if not results:
        lines.append("no results found")
        return "\n".join(lines)

    for idx, item in enumerate(results, 1):
        title = str(item.get("title", "")).strip() or "(untitled)"
        url = str(item.get("url", "")).strip()
        snippet = str(item.get("snippet", "")).strip()
        lines.append(f"{idx}. {title}")
        if url:
            lines.append(f"url: {url}")
        if snippet:
            lines.append(f"snippet: {snippet}")
    return "\n".join(lines)

File: gateway_unified/claude-gateway-rs/src/web_search.rs
#![allow(dead_code)]

use anyhow::Result;
use regex::Regex;
use std::time::Duration;
use url::Url;

lazy_static::lazy_static! {
    static ref RESULT_ANCHOR_RE: Regex = Regex::new(r#"(?is)<a[^>]*class="result__a"[^>]*href="([^"]+)"[^>]*>(.*?)</a>"#).unwrap();
    static ref RESULT_SNIPPET_RE: Regex = Regex::new(r#"(?is)class="result__snippet"[^>]*>(.*?)</(?:a|div)>"#).unwrap();
    static ref TAG_RE: Regex = Regex::new(r#"<[^>]+>"#).unwrap();
}

fn strip_tags(raw: &str) -> String {
    let text = TAG_RE.replace_all(raw, "");
    html_escape::decode_html_entities(&text).trim().to_string()
}

fn normalize_url(raw_href: &str) -> String {
    let mut value = html_escape::decode_html_entities(raw_href.trim()).into_owned();
    if value.is_empty() {
        return String::new();
    }

    if value.starts_with("//") {
        value = format!("https:{}", value);
    } else if value.starts_with("/l/?") {
        value = format!("https://duckduckgo.com{}", value);
    }

    if let Ok(parsed) = Url::parse(&value) {
        if parsed.host_str() == Some("duckduckgo.com") && parsed.path() == "/l/" {
            for (k, v) in parsed.query_pairs() {
                if k == "uddg" {
                    return urlencoding::decode(&v)
                        .map(|s| s.trim().to_string())
                        .unwrap_or_else(|_| v.trim().to_string());
                }
            }
        }
    }
    value
}

fn domain_allowed(url: &str, allowed_domains: Option<&[String]>) -> bool {
    let allowed = match allowed_domains {
        Some(domains) => domains,
        None => return true,
    };
    if allowed.is_empty() {
        return true;
    }

    if let Ok(parsed) = Url::parse(url) {
        if let Some(netloc) = parsed.host_str() {
            let netloc = netloc.to_lowercase();
            for domain in allowed {
                let d = domain.trim().to_lowercase();
                if d.is_empty() {
                    continue;
                }
                if netloc == d || netloc.ends_with(&format!(".{}", d)) {
                    return true;
                }
            }
        }
    }
    false
}

pub struct SearchResult {
    pub title: String,
    pub url: String,
    pub snippet: String,
}

async fn search_duckduckgo_html_once(
    query: &str,
    max_results: usize,
    timeout_s: f64,
    allowed_domains: Option<&[String]>,
) -> Result<Vec<SearchResult>> {
    let q = query.trim();
    if q.is_empty() {
        return Ok(Vec::new());
    }

    let client = reqwest::Client::builder()
        .timeout(Duration::from_secs_f64(timeout_s))
        .redirect(reqwest::redirect::Policy::limited(10))
        .build()?;

    let resp = client.get("https://html.duckduckgo.com/html/")
        .query(&[("q", q)])
        .header("user-agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/136.0.0.0 Safari/537.36")
        .header("accept-language", "zh-CN,zh;q=0.9,en;q=0.8")
        .send()
        .await?;

    if !resp.status().is_success() {
        return Ok(Vec::new());
    }

    let html_text = resp.text().await?;
    let mut results = Vec::new();

    for cap in RESULT_ANCHOR_RE.captures_iter(&html_text) {
        let href = normalize_url(cap.get(1).map(|m| m.as_str()).unwrap_or(""));
        if href.is_empty() || !domain_allowed(&href, allowed_domains) {
            continue;
        }

        let title = strip_tags(cap.get(2).map(|m| m.as_str()).unwrap_or(""));
        if title.is_empty() {
            continue;
        }

        let m = cap.get(0).unwrap();
        let end_idx = m.end();
        let tail_end = std::cmp::min(end_idx + 2200, html_text.len());
        let tail = &html_text[end_idx..tail_end];

        let mut snippet = String::new();
        if let Some(snippet_cap) = RESULT_SNIPPET_RE.captures(tail) {
            snippet = strip_tags(snippet_cap.get(1).map(|m| m.as_str()).unwrap_or(""));
        }

        results.push(SearchResult {
            title,
            url: href,
            snippet,
        });

        if results.len() >= max_results {
            break;
        }
    }

    Ok(results)
}

pub async fn search_duckduckgo_html(
    query: &str,
    max_results: usize,
    timeout_s: f64,
    allowed_domains: Option<&[String]>,
    retries: usize,
) -> Result<Vec<SearchResult>> {
    let mut last_err = None;
    let total_attempts = std::cmp::max(1, retries + 1);

    for attempt in 0..total_attempts {
        match search_duckduckgo_html_once(query, max_results, timeout_s, allowed_domains).await {
            Ok(res) => return Ok(res),
            Err(e) => {
                println!(
                    "[gateway web_search] retryable_error attempt={}/{} type={}",
                    attempt + 1,
                    total_attempts,
                    e
                );
                last_err = Some(e);
                if attempt >= total_attempts - 1 {
                    break;
                }
                tokio::time::sleep(Duration::from_millis(350)).await;
            }
        }
    }

    if let Some(e) = last_err {
        return Err(e);
    }
    Ok(Vec::new())
}

pub fn format_web_search_tool_result_text(query: &str, results: &[SearchResult]) -> String {
    let q = query.trim();
    let q_display = if q.is_empty() { "(empty)" } else { q };
    let mut lines = vec![format!("query: {}", q_display)];

    if results.is_empty() {
        lines.push("no results found".to_string());
        return lines.join("\n");
    }

    for (idx, item) in results.iter().enumerate() {
        let title = if item.title.trim().is_empty() {
            "(untitled)"
        } else {
            item.title.trim()
        };
        lines.push(format!("{}. {}", idx + 1, title));
        if !item.url.is_empty() {
            lines.push(format!("url: {}", item.url.trim()));
        }
        if !item.snippet.is_empty() {
            lines.push(format!("snippet: {}", item.snippet.trim()));
        }
    }

    lines.join("\n")
}
